### PR TITLE
Backport contextual commands from 5.0 to 4.8.x in order to use it in Artik IDE

### PR DESCRIPTION
**Pull request by azatsarynnyy**
### What does this PR do?

Backport contextual commands from 5.0 to 4.8.x in order to use it in Artik IDE.
Add `${explorer.current.file.parent.path}` macro.
Add two attributes for C PT: `binaryName` and `compilationOptions`.
### What issues does this PR fix or reference?

It's needed for https://github.com/codenvy/artik-ide/issues/139.

@vparfonov

### CODE REVIEW

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/part/explorer/project/macro/ExplorerCurrentFileParentPathProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @Test
+    public void getEmptyValues() throws Exception {
+        initWithNoFiles();
+
+        provider.getValue();
+
+        verify(promiseProvider).resolve(eq(""));
+    }
+
+}
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/part/explorer/project/macro/ExplorerCurrentFileParentPathProviderTest.java]: new line

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionFactory.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Factory for creating {@link CommandProducerAction} instances.
+ *
+ * @author Artem Zatsarynnyi
+ */
+public interface CommandProducerActionFactory {
+
+    /** Creates action for executing command produced by the specified {@code commandProducer}. */
+    CommandProducerAction create(String name, CommandProducer commandProducer, Machine machine);
+}
```
Comment (skabashnyuk) [inline plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionFactory.java]: new line

Comment (azatsarynnyy) [reply plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionFactory.java]: thx. Fixed

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        for (CommandProducer commandProducer : commandProducers) {
+            createActionsForProducer(commandProducer);
+        }
+    }
+
+    @Override
+    public void onWsAgentStopped(WsAgentStateEvent event) {
+    }
+
+    private void createActionsForProducer(CommandProducer producer) {
```
Comment (vparfonov) [inline plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:142]: Pls, provide javadoc here

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+            for (Machine machine : machines) {
+                createActionsForMachine(machine);
+            }
+        }
+
+        commandProducersActionsGroup.add(action);
+    }
+
+    private void createActionsForMachine(Machine machine) {
```
Comment (vparfonov) [inline plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:168]: javadoc, with some example

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.eclipse.che.ide.api.action.IdeActions.GROUP_EDITOR_TAB_CONTEXT_MENU;
+import static org.eclipse.che.ide.api.action.IdeActions.GROUP_MAIN_CONTEXT_MENU;
+
+/**
+ * Manages actions for contextual commands.
+ *
```
Comment (vparfonov) [inline plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:46]: more info here, look like this major class for this feature, must be described more detail. Maybe with some use case that can be solved with it

Comment (azatsarynnyy) [reply plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:46]: added more detailed javadoc including explanation how component works

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java (hunk 4 above)
Comment (azatsarynnyy) [reply plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:168]: added javadoc

File: plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java (hunk 3 above)
Comment (azatsarynnyy) [reply plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/producer/CommandProducerActionManager.java:142]: added javadoc

**Comment by azatsarynnyy** [comment]
@vparfonov I've fixed notes

**Review by vparfonov — APPROVED** [review]

**Pull request merged**